pkg/foundation/semaphore/semaphore.go

```diff
@@ -16,71 +16,102 @@ package semaphore
+	// previous lock is released, the value i will be sent into ready, signaling
+	// this ticket that it can acquire a Lock.
+	ready chan int
+}
+
+// Lock represents a lock of the semaphore. After the lock is not needed any
+// more it should be released.
+type Lock struct {
+	// i is the index of the current lock.
+	i int
+	// next is the same channel as Ticket.ready of the next ticket. Once Lock is
+	// released, the value i+1 will be sent into next, signaling the next ticket
+	// that it can acquire a Lock.
+	next chan int
 }
 
 // Enqueue reserves the next place in the queue and returns a Ticket used to
-// acquire access to the resource when it's the callers turn. The Ticket has to
-// be supplied to Release before discarding.
+// acquire access to the Lock when it's the callers turn. The Ticket has to be
+// supplied to Acquire and discarded afterwards.
+// Enqueue is not safe for concurrent use.
 func (s *Simple) Enqueue() Ticket {
-	s.mu.Lock()
-	defer s.mu.Unlock()
+	s.init.Do(func() {
+		s.chanPool.New = func() any {
+			return make(chan int, 1) // make it buffered to not block the release of a lock
+		}
+
+		s.last.next = s.chanPool.Get().(chan int)
+		s.last.next <- s.last.i + 1 // first lock can be acquired right away
+	})
 
 	t := Ticket{
-		ready: s.lastTicket.next,
-		next:  make(chan struct{}),
-	}
-	if t.ready == nil {
-		// first time we create a ticket it will be already acquired
-		t.ready = make(chan struct{})
-		close(t.ready)
+		Lock: Lock{
+			i:    s.last.i + 1,
+			next: s.chanPool.Get().(chan int),
+		},
+		ready: s.last.next,
 	}
-	s.lastTicket = t
+	s.last = t.Lock
 	return t
 }
 
-// Acquire acquires the semaphore, blocking until resources are available.
-// Returns nil if acquire was successful or ctx.Err if the context was cancelled
-// in the meantime.
-func (s *Simple) Acquire(t Ticket) {
-	<-t.ready
-}
-
-// Release releases the semaphore and notifies the next in line if any.
-// If the ticket was already released the function returns an error. After the
-// ticket is released it should be discarded.
-func (s *Simple) Release(t Ticket) error {
-	select {
-	case <-t.ready:
-	default:
-		return cerrors.New("semaphore: can't release ticket that was not acquired")
+// Acquire acquires a Lock, blocking until the previous ticket is released.
+// Ticket needs to be discarded after the call to Acquire. The Lock has to be
+// supplied to Release after it is not needed anymore and discarded afterwards.
+// Acquire is safe for concurrent use.
+func (s *Simple) Acquire(t Ticket) Lock {
+	i := <-t.ready
+	if t.i != i {
+		// Multiple reasons this can happen. A ticket other than this one could
+		// have been successfully released twice. The other possibility is that
+		// this ticket is trying to be acquired after it was released. Both
+		// cases are invalid.
+		s.panic()
 	}
+	s.chanPool.Put(t.ready)
+	return t.Lock // return only the lock part of the ticket
```

Comment:
It's mostly to make the interface clearer - previously the same `Ticket` had to be used first in `Acquire` and then in `Release`, so theoretically someone could think they can just skip `Acquire` and directly `Release` a ticket. This wasn't actually a valid use of the semaphore. This potential misuse is now prevented by the compiler since you need to call `Acquire` to get a `Lock` that can then be provided to `Release`.

Reply:
That makes sense. :+1: